geomstats/symmetric_positive_definite_matrices.py

```diff
@@ -0,0 +1,216 @@
+"""
+Computations on the manifold of
+symmetric positive definite matrices.
+
+X. Pennec. A Riemannian Framework for Tensor Computing. (2004).
+"""
+
+import logging
+import math
+import numpy as np
+import scipy.linalg
+
+TOL_SYMMETRIC = 1e-12
+EPSILON = 1e-6
+
+
+def is_symmetric(mat, tolerance=TOL_SYMMETRIC):
+    """
+    Check if a matrix is symmetric.
+    """
+    return np.allclose(mat, mat.transpose(), atol=tolerance)
+
+
+def belongs(mat, tolerance=TOL_SYMMETRIC):
+    """
+    Check if a matrix belong to the manifold of
+    symmetric positive definite matrices.
+    """
+    if is_symmetric(mat, tolerance=tolerance):
+        eigenvalues = np.linalg.eigvalsh(mat)
+        return np.all(eigenvalues > 0)
+    return False
+
+
+# TODO(nina): The manifold of sym matrices is not a Lie group.
+# Use 'group_exp' and 'group_log'?
+def group_exp(sym_mat):
+    """
+    Group exponential of the Lie group of
+    all invertible matrices has a straight-forward
+    computation for symmetric positive definite matrices.
+    """
+    assert is_symmetric(sym_mat)
+    # Make it fully symmetric for numerics.
+    sym_mat = (sym_mat + sym_mat.transpose()) / 2
+
+    [eigenvalues, vectors] = np.linalg.eigh(sym_mat)
+
+    diag_exp = np.diag(np.exp(eigenvalues))
+    exp = np.dot(np.dot(vectors, diag_exp), vectors.transpose())
+
+    return(exp)
+
+
+def group_log(sym_mat):
+    """
+    Group logarithm of the Lie group of
+    all invertible matrices has a straight-forward
+    computation for symmetric positive definite matrices.
+    """
+    assert is_symmetric(sym_mat)
+    # Make it fully symmetric for numerics.
+    sym_mat = (sym_mat + sym_mat.transpose()) / 2
+
+    [eigenvalues, vectors] = np.linalg.eigh(sym_mat)
+
+    assert np.all(eigenvalues > 0)
+
+    diag_log = np.diag(np.log(eigenvalues))
+    log = np.dot(np.dot(vectors, diag_log), vectors.transpose())
+
+    return log
+
+
+def matrix_to_vector(matrix):
+    """
+    Convert the symmetric part of a symmetric matrix
+    into a vector.
+    """
+    assert is_symmetric(matrix)
+
+    dim_mat, _ = matrix.shape
+    dim_vec = dim_mat * (dim_mat + 1) / 2
```

Comment:
be careful here, while this might work with python2, python3 might behave differently with float. Make sure to always force the behavior you expect wrt integers.